keras_cv/layers/regularization/dropblock_2d.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+from keras.engine.base_layer import BaseRandomLayer
+from keras.utils import conv_utils
+
+
+class DropBlock2D(BaseRandomLayer):
+    """Applies DropBlock regularization to input features.
+
+    DropBlock is a form of structured dropout, where units in a contiguous
+    region of a feature map are dropped together. DropBlock works better than
+    dropout on convolutional layers due to the fact that activation units in
+    convolutional layers are spatially correlated.
+
+    It is advised to use DropBlock after activation in Conv -> BatchNorm -> Activation
+    block in further layers of the network. For example, the paper mentions using
+    DropBlock in 3rd and 4th group of ResNet blocks.
+
+    Reference:
+    - [DropBlock: A regularization method for convolutional networks](
+        https://arxiv.org/abs/1810.12890
+    )
+
+    Args:
+        keep_probability: float. Probability of keeping a unit. Defaults to 0.9.
```

Comment:
We might need to make this a tf.variable internally so it doesn't get baked into the graph to support a keep_probability schedule callback.  Not sure yet, no action item needed on your side.  Still doing some digging here.

Reply:
Yes I was also thinking about the Schedule Callback. Something like
```python
class DropBlockScheduleCallback(tf.keras.callbacks.Callback):
    def __init__(self, total_steps: int):
        super(DropBlockScheduleCallback, self).__init__()
        self._total_steps = total_steps
        self._current_step = 0

    def on_train_batch_end(self):
        self._current_step += 1
        for layer in self.model.layers:
            if isinstance(layer, DropBlock2D):
                new_keep_probability = self._calc_new_keep_probability(
                    layer._keep_probability  # TODO: public
                )
                layer._keep_probability.assign(new_keep_probability)

    def _calc_new_keep_probability(self, current_keep_probability):
        current_ratio = self._current_step / self._total_steps
        return 1 - current_ratio * (1 - current_keep_probability)
```

Although I'm not sure:
1. is there any other way to get `total_steps` without having them need to be explicitly provided.
2. is iterating over all model layers for each batch not too slow. Maybe they could be saved `on_train_begin` somewhere and used as a reference `on_batch_end`?

If I'm thinking in the right direction, I could add this in this PR. Or this could be a separate issue. Overall, please let me know.

Reply:
I don't know if this is easy to implement, but it would be very cool if could accept lambdas or Schedules:

```py
global_step = tf.Variable(0, trainable=False)
db = DropBlock(keep_probability=lambda: 1/step)
# or ...
first_keep_prob = .9
db = DropBlock(
  keep_probability=tf.keras.optimizers.schedules.InverseTimeDecay(
    1.,
    decay_steps=steps_per_epoch*epochs,
    decay_rate=0.96))
```

Reply:
Yeah, I'm not sure what would be optimal solution for this parameter's scheduling. I would probably add it in a separate PR (+ separate issue), as there is still some work regarding the basic functionality.

Reply:
I'll discuss this with the team and let you know.  It would be good to get this sorted out because we actually will have the same problem with a few other components.